### ISSUE TITLE
CloudStorageFileSystem: ConcurrentModificationException belies thread safety

### ISSUE THREAD

**Issue by cjllanwarne** [googleapis/java-storage-nio#691]
We recently updated from an older version of the storage NIO library and are now seeing `ConcurrentModificationExceptions` from `CloudStorageFileSystem`. At a guess, it seems like this function is perhaps not perfectly resilient if `CONFIG_TO_PROVIDERS_MAP` is updated while the `existingProviders` value is being iterated over?

```
Caused by: java.util.ConcurrentModificationException: null
	at java.base/java.util.HashMap$HashIterator.nextNode(Unknown Source)
	at java.base/java.util.HashMap$KeyIterator.next(Unknown Source)
	at com.google.cloud.storage.contrib.nio.CloudStorageFileSystem.getCloudStorageFileSystemProvider(CloudStorageFileSystem.java:164)
	at com.google.cloud.storage.contrib.nio.CloudStorageFileSystem.forBucket(CloudStorageFileSystem.java:196)
	at [...our client code]
```

**Comment by breilly2** [comment]
https://github.com/googleapis/java-storage-nio/blob/7f6f65d1b29e289006472e2a1e9a76cfa495eeec/google-cloud-nio/src/main/java/com/google/cloud/storage/contrib/nio/CloudStorageFileSystem.java#L154 Looks very problematic:

```java
  private static CloudStorageFileSystemProvider getCloudStorageFileSystemProvider(
      CloudStorageConfiguration config, StorageOptions storageOptions) {
    CloudStorageFileSystemProvider newProvider =
        (storageOptions == null)
            ? new CloudStorageFileSystemProvider(config.userProject())
            : new CloudStorageFileSystemProvider(config.userProject(), storageOptions);
    Set<CloudStorageFileSystemProvider> existingProviders = CONFIG_TO_PROVIDERS_MAP.get(config);
    if (existingProviders == null) {
      existingProviders = new HashSet<>();
    } else {
      for (CloudStorageFileSystemProvider existiningProvider : existingProviders) {
        if (existiningProvider.equals(newProvider)) {
          return existiningProvider;
        }
      }
    }
    existingProviders.add(newProvider);
    CONFIG_TO_PROVIDERS_MAP.put(config, existingProviders);
    return newProvider;
  }
```
1. `CloudStorageFileSystem` is annotated as `@ThreadSafe`. Perhaps its instances are, but this static function is not. `existingProviders` is a `HashSet` which gets modified near the end of the function without any synchronization. Concurrent callers with the same `config` and `storageOptions` can run into this `ConcurrentModificationException`.
2. The values of `existingProviders` are `CloudStorageFileSystemProvider` instances (which are oddly annotated as `@Singleton`). `CloudStorageFileSystemProvider`'s `equals` and `hashCode` are based on the `storage` property which is modified after object instantiation (from `initStorage` called from various places). Therefore, once a "cached" instance has been initialized, `existingProvider.equals(newProvider)` will yield `false` (since `newProvider` hasn't yet been initialized).

**Comment by JesseLovelace** [comment]
Thanks for the feedback, we're taking a look at this. How often does this issue come up for you?

**Comment by aednichols** [comment]
The issue reproduces more or less continuously when our multitenant server app is in production with the affected release.

We also created a [unit test](https://github.com/broadinstitute/cromwell/pull/6491/files#diff-758dbbe823e71cc26fee7bc89cd5c434dfb76e604d51005b8327db59aab96068R300-R336) that triggers it in a minimal scenario.

**Comment by JesseLovelace** [comment]
Thanks! We'll have an update on this tomorrow

**Comment by aednichols** [comment]
Sounds great, thank you!

**Comment by BenWhitehead** [comment]
Thank you very much for providing the minimised repo test that helped a lot.

I've proposed a fix in #719 which fixes the thread safety issue, and fixes the actual caching mapping outlined in #698 as well.

**Comment by aednichols** [comment]
Excellent, thank you very much! Appreciate the fast turnaround.
